src/python/deepgnn/graph_engine/snark/converter/writers.py

```diff
@@ -17,16 +16,97 @@
 from deepgnn.graph_engine.snark.meta import _Element
 
 
-class NodeWriter:
-    """
-    NodeWriter is an entry point for conversion from euler json files to DeepGNN files.
+class BinaryWriter:
+    """BinaryWriter is an entry point for conversion from input files to binary files.
 
     Every record is parsed one by one and passed to the relevant writers below. These writers
     put data in files with the partition suffix: `node_{partition}...`. We reserve one more
     suffix to follow after partition to split large files into smaller ones, so files will have names
     like this: `node_{partition}_{iteration}...`.
     """
 
+    def __init__(
+        self,
+        folder: str,
+        suffix: int,
+        node_type_num: int,
+        edge_type_num: int,
+        skip_node_sampler: bool,
+        skip_edge_sampler: bool,
+    ):
+        """Initialize writer and create binary files.
+
+        Args:
+            folder (str): where to save binaries
+            suffix (int): file suffix in the name of binary files
+            node_type_num (int): number of node types in the graph
+            edge_type_num (int): number of edge types in the graph
+            skip_node_sampler(bool): skip generation of node alias tables
+            skip_edge_sampler(bool): skip generation of edge alias tables
+        """
+        self.folder = folder
+        self.suffix = suffix
+        self.node_type_num = node_type_num
+        self.edge_type_num = edge_type_num
+        self.skip_node_sampler = skip_node_sampler
+        self.skip_edge_sampler = skip_edge_sampler
+
+        self.node_writer = NodeWriter(self.folder, self.suffix)
+        self.edge_writer = EdgeWriter(self.folder, self.suffix)
+
+        self.node_alias: typing.Union[NodeAliasWriter, _NoOpWriter] = (
+            _NoOpWriter()
+            if skip_node_sampler
+            else NodeAliasWriter(self.folder, self.suffix, self.node_type_num)
+        )
+        self.edge_alias: typing.Union[EdgeAliasWriter, _NoOpWriter] = (
+            _NoOpWriter()
+            if skip_edge_sampler
+            else EdgeAliasWriter(self.folder, self.suffix, self.node_type_num)
+        )
+
+        self.node_count = 0
+        self.edge_count = 0
+        self.node_weight = [0.0] * self.node_type_num
+        self.node_type_count = [0] * self.node_type_num
+        self.edge_weight = [0.0] * self.edge_type_num
+        self.edge_type_count = [0] * self.edge_type_num
+
+    def add(self, data: typing.Iterator[typing.Tuple[int, int, int, float, list]]):
+        """
+        Write binary for a node and all of its edges.
+
+        args:
+            data: Iterable[(int, int, int, float, list)] Data for a node first, then all of
+                its edges in order of dst. Each entry for a node/edge is,
+                (-1/src, node_id/dst, type, weight, [ndarray for each feature vector or None in order of feature index]).
```

Comment:
how would you sort a following graph file with two nodes 2 and 3 connected to each other?
```csv
-1 2 0 1 [1,1,1]
-1 3 0 1 [2,2,2]
 2 3 0 1 []
 3 2 0 1 []
 ```
 It seems like it requires some manual reordering and for large graphs I can't think of any sorting procedure that will organize graph items in the required order.

Reply:
Right now,
```
  -1 2 0 1 [1,1,1]
  2 3 0 1 []
  -1 3 0 1 [2,2,2]
  3 2 0 1 []
```
In the future when nodes/edges do not need to be on the same partition it would not need to change.

Reply:
I could write a function from this file to a buffer file that would be the one to get converted, if the original file is sorted all nodes then all edges the complexity would not be bad.